### PR TITLE
Shooter

### LINKED ISSUE #86: Shooter
Tracking issue for the shooter (double flywheel so Wom in built won't be used)

### PR DESCRIPTION
**Pull request by totallysomeoneyoudontknow**
**What this PR does**
Implements shooter

**What does it fix/add?**
Resolves #86

### CODE REVIEW

File: src/main/cpp/Robot.cpp [hunk 1]
```diff
@@ -1,11 +1,33 @@
 // Copyright (c) 2023-2024 CurtinFRC
 // Open Source Software, you can modify it according to the terms
 // of the MIT License at the root of this project
-
+// Copyright (c) 2023-2024 CurtinFRC
```
Comment (goanna247) [inline src/main/cpp/Robot.cpp:4]: Why do you have the licence twice

File: src/main/cpp/Shooter.cpp [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+
+void Shooter::OnUpdate(units::second_t dt){
+    switch(_state){
+        case ShooterState::kIdle:
+				{
+          _config.ShooterGearbox.transmission->SetVoltage(0_V);
+          if (_shooterSensor.Get()) {       
+            _state = ShooterState::kReverse;
+            // kSpinup
```
Comment (goanna247) [inline src/main/cpp/Shooter.cpp:16]: remove comments

File: src/main/cpp/Shooter.cpp [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+            // kShooting
+          } 
+				}
+        break;
+        case ShooterState::kSpinUp:
+        {
+          _pid.SetSetpoint(_goal);
+          _config.ShooterGearbox.transmission->SetVoltage(0.2_V);
```
Comment (goanna247) [inline src/main/cpp/Shooter.cpp:24]: you need you .Calculate pid loop here

File: src/main/cpp/Shooter.cpp [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+          if (_pid.IsStable()) {
+            setState(ShooterState::kShooting);
+          }
+
+        }
+        break;
+        case ShooterState::kShooting:
+				{
+          _pid.SetSetpoint(_goal);
+        	_config.ShooterGearbox.transmission->SetVoltage(0.2_V);
```
Comment (goanna247) [inline src/main/cpp/Shooter.cpp:35]: the voltage set needs to be the output of a pid loop

File: src/main/cpp/Shooter.cpp [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+          // if (_pid.IsStable()) {
```
Comment (goanna247) [inline src/main/cpp/Shooter.cpp:36]: if the pid loop is not stable then we should move back to the spinup state

File: src/main/cpp/Shooter.cpp [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+            // setState(ShooterState::kIdle);
```
Comment (goanna247) [inline src/main/cpp/Shooter.cpp:37]: remove comments

File: src/main/cpp/Shooter.cpp [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+          // do we need this first loop?
+          if (_shooterSensor.Get()) {
+            setState(ShooterState::kIdle);
+          }
+				}
+        break;
+        case ShooterState::kReverse:
+				{
+        	_config.ShooterGearbox.transmission->SetVoltage(-5_V);
+          if (_shooterSensor.Get()) {
```
Comment (goanna247) [inline src/main/cpp/Shooter.cpp:48]: if shooterSensor is NOT tripped

File: src/main/cpp/Shooter.cpp [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+            setState(ShooterState::kIdle);
+          }
+				}
+        break;
+        case ShooterState::kRaw:
+				{
+        	_config.ShooterGearbox.transmission->SetVoltage(5_V);
```
Comment (goanna247) [inline src/main/cpp/Shooter.cpp:55]: this should be the _rawVoltage being set to the shooter here

File: src/main/cpp/Shooter.cpp [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+          if (_shooterSensor.Get()) {
```
Comment (goanna247) [inline src/main/cpp/Shooter.cpp:56]: this loop does nothing, you probably dont want it

File: src/main/cpp/ShooterBehaviour.cpp [hunk 10]
```diff
@@ -0,0 +1,38 @@
+#include "ShooterBehaviour.h"
+
+
+ShooterManualControl::ShooterManualControl(Shooter *shooter, frc::XboxController &codriver): _shooter(shooter), _codriver(codriver){
+    Controls(shooter);
+}
+
+void ShooterManualControl::OnTick(units::second_t dt) {
+    if (_codriver.GetLeftBumper()) {
```
Comment (goanna247) [inline src/main/cpp/ShooterBehaviour.cpp:9]: you dont need this stuff you already have it below

File: src/main/cpp/ShooterBehaviour.cpp [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+       _shooter->setState(ShooterState::kSpinUp);
+       _shooter->setPidGoal(65_rad_per_s);
+    } else if (_codriver.GetRightBumper()) {
+       _shooter->setState(ShooterState::kSpinUp);
+       _shooter->setPidGoal(20_rad_per_s);
+    }
+    
+    double rawControl;
```
Comment (goanna247) [inline src/main/cpp/ShooterBehaviour.cpp:17]: this should be a boolean stored in the .h file

File: src/main/cpp/ShooterBehaviour.cpp [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+    if(_codriver.GetAButtonPressed()){
+        if (rawControl == true){
+            rawControl = false;
+        } else {
+            rawControl = true;
+        }
+    }
+if (!rawControl) { // rawControl == false
```
Comment (goanna247) [inline src/main/cpp/ShooterBehaviour.cpp:25]: fix your indenting

File: src/main/cpp/ShooterBehaviour.cpp [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+    if (_codriver.GetLeftBumper()) {
+        _shooter->setState(ShooterState::kSpinUp);
+        _shooter->setPidGoal(65_rad_per_s);
+    } else if (_codriver.GetRightBumper()) {
+        _shooter->setState(ShooterState::kSpinUp);
+        _shooter->setPidGoal(20_rad_per_s);
+    }
+} else {
+     _shooter->setRaw(0_V);
```
Comment (goanna247) [inline src/main/cpp/ShooterBehaviour.cpp:34]: if the shooter is in raw you want to map the set raw to a trigger input

File: src/main/include/RobotMap.h [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+struct RobotMap {
+    frc::XboxController driver{0};
+    frc::XboxController codriver{1};
+
+    struct Shooter {
+        rev::CANSparkMax shooterMotor{99, rev::CANSparkMax::MotorType::kBrushless};
+        frc::DigitalInput shooterSensor{2};
+
+        wom::VoltageController shooterMotorGroup = wom::VoltageController::Group(shooterMotor);
+        wom::CANSparkMaxEncoder shooterEncoder=wom::CANSparkMaxEncoder(&shooterMotor, 1);
```
Comment (goanna247) [inline src/main/include/RobotMap.h:15]: pretty sure the second number here is ticks, which should be a different value

**Review by goanna247 — CHANGES_REQUESTED** [review]

File: src/main/cpp/Shooter.cpp [hunk 15]
```diff
@@ -13,15 +13,13 @@ void Shooter::OnUpdate(units::second_t dt){
...
-            // kSpinup
-            // kShooting
           } 
 				}
         break;
         case ShooterState::kSpinUp:
         {
           _pid.SetSetpoint(_goal);
-          _config.ShooterGearbox.transmission->SetVoltage(0.2_V);
+          _config.ShooterGearbox.transmission->SetVoltage(_pid.Calculate(_config.ShooterGearbox.encoder->GetEncoderAngularVelocity(), dt));
```
Comment (goanna247) [inline src/main/cpp/Shooter.cpp:22]: put the Calculate function on a new line and assign it to a variable

File: src/main/cpp/Shooter.cpp [hunk 16]
```diff
@@ -32,11 +30,11 @@ void Shooter::OnUpdate(units::second_t dt){
         case ShooterState::kShooting:
 				{
           _pid.SetSetpoint(_goal);
-        	_config.ShooterGearbox.transmission->SetVoltage(0.2_V);
-          // if (_pid.IsStable()) {
-            // setState(ShooterState::kIdle);
-          // }
-          // do we need this first loop?
+        	_config.ShooterGearbox.transmission->SetVoltage(_pid.Calculate(_config.ShooterGearbox.encoder->GetEncoderAngularVelocity(), dt));
```
Comment (goanna247) [inline src/main/cpp/Shooter.cpp:33]: put the calculate function on a new line and assign it to a variable

File: src/main/cpp/Shooter.cpp [hunk 17]
```diff
@@ -32,11 +30,11 @@ void Shooter::OnUpdate(units::second_t dt){
...
+
+           if (_pid.IsStable()) {
```
Comment (goanna247) [inline src/main/cpp/Shooter.cpp:35]: if PID is not stable

File: src/main/cpp/ShooterBehaviour.cpp [hunk 18]
```diff
@@ -6,33 +6,26 @@ ShooterManualControl::ShooterManualControl(Shooter *shooter, frc::XboxController
...
+    if (!_rawControl) { // rawControl == false
+        if (_codriver.GetLeftBumper()) {
+            _shooter->setState(ShooterState::kSpinUp);
+            _shooter->setPidGoal(65_rad_per_s);
+        } else if (_codriver.GetRightBumper()) {
+            _shooter->setState(ShooterState::kSpinUp);
+            _shooter->setPidGoal(20_rad_per_s);
+        }
+    } else {
+         _shooter->setRaw(_codriver.GetRightBumperPressed() * 12_V);
```
Comment (goanna247) [inline src/main/cpp/ShooterBehaviour.cpp:27]: bumper is a button which means it either returns a true or a false, you should use something that returns an analog value (something that has 0-1 values) such as a trigger or a joystick

File: src/main/include/ShooterBehaviour.h [hunk 19]
```diff
@@ -15,4 +15,7 @@ class ShooterManualControl : public behaviour::Behaviour{
   Shooter* _shooter;
   frc::XboxController &_codriver;
 
+  bool _rawControl;
```
Comment (goanna247) [inline src/main/include/ShooterBehaviour.h:18]: set a default for this, either true or false

**Review by goanna247 — CHANGES_REQUESTED** [review]

File: src/main/include/Shooter.h [hunk 20]
```diff
@@ -0,0 +1,41 @@
...
+
+
+class Shooter : public behaviour::HasBehaviour{
+  public:
+    Shooter(ShooterConfig config);
+
+    void OnUpdate(units::second_t dt);
+    void setState(ShooterState state);
+    void setRaw(units::volt_t voltage);
+    void setPidGoal(units::radians_per_second_t);
```
Comment (spacey-sooty) [inline src/main/include/Shooter.h:30]: Function names should start with a capital

File: src/main/cpp/ShooterBehaviour.cpp [hunk 21]
```diff
@@ -0,0 +1,31 @@
...
+
+
+    if(_codriver.GetAButtonPressed()){
+        if (_rawControl == true){
+            _rawControl = false;
+        } else {
+            _rawControl = true;
+        }
+    }
+    if (!_rawControl) { // rawControl == false
```
Comment (spacey-sooty) [inline src/main/cpp/ShooterBehaviour.cpp:18]: remove comment

File: src/main/cpp/ShooterBehaviour.cpp [hunk 22]
```diff
@@ -0,0 +1,31 @@
...
+
+void ShooterManualControl::OnTick(units::second_t dt) {
+
+
+    if(_codriver.GetAButtonPressed()){
+        if (_rawControl == true){
+            _rawControl = false;
+        } else {
+            _rawControl = true;
+        }
```
Comment (spacey-sooty) [inline src/main/cpp/ShooterBehaviour.cpp:16]: you can do _rawControl != _rawControl

**Review by spacey-sooty — CHANGES_REQUESTED** [review]

**Pull request closed without merging**